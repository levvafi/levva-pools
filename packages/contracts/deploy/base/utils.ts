import { ethers } from 'ethers';

export function validateAddress(address: string) {
  if (!ethers.isAddress(address) || address === ethers.ZeroAddress) {
    throw new Error(`Wrong ${address} evm address`);
  }
}
