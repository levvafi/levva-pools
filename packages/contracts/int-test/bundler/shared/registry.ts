export const WETH9 = '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2';
export const WSTETH = '0x7f39c581f595b53c5cb19bd0b3f8da6c935e2ca0';
export const USDC = '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48';

export const PENDLE_ROUTER = '0x888888888889758F76e7103c6CbF23ABbF58F946';
export const PENDLE_ORACLE = '0x9a9fa8338dd5e5b2188006f1cd2ef26d921650c2';

export const PENDLE_PT_WSTETH_30_12_27 = '0xb253eff1104802b97ac7e3ac9fdd73aece295a2c';
export const PENDLE_MARKET_WSTETH_30_12_27 = '0x34280882267ffa6383b363e278b027be083bbe3b';
