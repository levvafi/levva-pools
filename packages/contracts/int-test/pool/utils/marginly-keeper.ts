import { AbiCoder, Addressable } from 'ethers';

export function encodeLiquidationParams(
  asset: string | Addressable,
  amount: bigint,
  marginlyPool: string | Addressable,
  positionToLiquidate: string | Addressable,
  liquidator: string | Addressable,
  uniswapPool: string | Addressable,
  minProfit: bigint,
  swapCallData: bigint
): string {
  /**
       *  address asset;
        uint256 amount;
        address marginlyPool;
        address positionToLiquidate;
        address liquidator;
        address uniswapPool;
        uint256 minProfit;
        uint256 swapCallData;
       */

  return AbiCoder.defaultAbiCoder().encode(
    ['address', 'uint256', 'address', 'address', 'address', 'address', 'uint256', 'uint256'],
    [asset, amount, marginlyPool, positionToLiquidate, liquidator, uniswapPool, minProfit, swapCallData]
  );
}
export function encodeLiquidationParamsAave(
  marginlyPool: string | Addressable,
  positionToLiquidate: string | Addressable,
  liquidator: string | Addressable,
  minProfit: bigint,
  swapCallData: bigint
): string {
  /**
    address marginlyPool;
    address positionToLiquidate;
    address liquidator;
    uint256 minProfit;
    uint256 swapCallData;
   */

  return AbiCoder.defaultAbiCoder().encode(
    ['address', 'address', 'address', 'uint256', 'uint256'],
    [marginlyPool, positionToLiquidate, liquidator, minProfit, swapCallData]
  );
}

export function encodeLiquidationParamsBalancer(
  marginlyPool: string | Addressable,
  positionToLiquidate: string | Addressable,
  liquidator: string | Addressable,
  minProfit: bigint,
  swapCallData: bigint
): string {
  /**
 *  
  address marginlyPool;
  address positionToLiquidate;
  address liquidator;
  uint256 minProfit;
  uint256 swapCallData;
 */

  return AbiCoder.defaultAbiCoder().encode(
    ['address', 'address', 'address', 'uint256', 'uint256'],
    [marginlyPool, positionToLiquidate, liquidator, minProfit, swapCallData]
  );
}
