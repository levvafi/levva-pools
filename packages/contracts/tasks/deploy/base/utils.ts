import { ethers } from 'ethers';

export function validateAddress(address: string) {
  if (!ethers.isAddress(address) || address === ethers.ZeroAddress) {
    throw new Error(`Wrong ${address} evm address`);
  }
}

export function isSameAddress(addressA: string, addressB: string): boolean {
  return ethers.getAddress(addressA) === ethers.getAddress(addressB);
}
