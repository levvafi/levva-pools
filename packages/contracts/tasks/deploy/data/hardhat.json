{
  "factory": {
    "poolType": "Trading",
    "feeHolderAddress": "0xd6DE5D2D8Ab2E4a92c2B923a88f7D366ddb3C760",
    "techPositionOwnerAddress": "0xd6DE5D2D8Ab2E4a92c2B923a88f7D366ddb3C760",
    "WETH9": {
      "address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
      "decimals": 18,
      "symbol": "WETH"
    }
  },
  "pools": [
    {
      "priceOracle": "0xd5Ea994A033737f0D568A2A307c65f8048b10698",
      "baseToken": {
        "address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
        "decimals": 18,
        "symbol": "WETH"
      },
      "quoteToken": {
        "address": "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
        "decimals": 6,
        "symbol": "USDC"
      },
      "defaultSwapCallData": 0,
      "params": {
        "maxLeverage": 20,
        "interestRate": 50000,
        "fee": 20000,
        "swapFee": 500,
        "mcSlippage": 50000,
        "positionMinAmount": 1e17,
        "quoteLimit": 100000e6
      }
    }
  ],
  "oracles": {
    "UniswapV3TickOracle": {
      "factoryAddress": "0x1F98431c8aD98523631AE4a59f267346ea31F984",
      "settings": [
        {
          "quoteToken": {
            "address": "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
            "decimals": 6,
            "symbol": "USDC"
          },
          "baseToken": {
            "address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
            "decimals": 18,
            "symbol": "WETH"
          },
          "uniswapFee": 500,
          "secondsAgo": 1800,
          "secondsAgoLiquidation": 5
        }
      ]
    }
  },
  "adapters": {
    "UniswapV3Adapter": {
      "dexId": 0,
      "settings": [
        {
          "token0": {
            "address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
            "decimals": 18,
            "symbol": "WETH"
          },
          "token1": {
            "address": "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
            "decimals": 6,
            "symbol": "USDC"
          },
          "poolAddress": "0x88e6A0c2dDD26FEEb64F039a2c41296FcB3f5640"
        }
      ]
    }
  }
}
