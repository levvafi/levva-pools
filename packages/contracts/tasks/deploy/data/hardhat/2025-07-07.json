{
  "LevvaTradingPoolImplementation": {
    "address": "0x43E9AC257b061919b001eb5eB90212F28F4A8901",
    "txHash": "0xe321485a0d1b428de4ef9675608920f58597477965fa42473c834e2d219f6414"
  },
  "MarginlyRouter": {
    "address": "0xD9cA22646bDbB0772ff1B15CC139e229f0AA7b64",
    "txHash": "0x1db8f32c15e42299c5c798971cb68935c76c70e5d9d4d9697b40198519921646"
  },
  "MarginlyFactory": {
    "address": "0xa4761E911dDC47f290928c0d3F7a908132289Ce7",
    "txHash": "0xb2d03c6ea93ddb3483f99fbfce6bbac4d4b0f6f0b87396341dcd5e4a3861cba0"
  },
  "UniswapV3TickOracle": {
    "address": "0xd5Ea994A033737f0D568A2A307c65f8048b10698",
    "txHash": "0x7da49ebe0353a3462095ec50f7b591c345d4cb92cf0309395df3897b6aa3f2d5"
  },
  "UniswapV3Adapter": {
    "address": "0x0Ee29547347E41fD7Bf14a0d064421275C2ae0bA",
    "txHash": "0xa4fd7d065356e21a7630d8b5b7c635dc14c94d621facec632c1c6d6554970906"
  },
  "LevvaPool#WETH-USDC": {
    "address": "0x66a129CBee62f674eC8461cba13398bf70e8334b",
    "txHash": "0x8e99990b8b444823e9cc88572db86246788c5e23b669512b68e29186ffda2462"
  }
}
