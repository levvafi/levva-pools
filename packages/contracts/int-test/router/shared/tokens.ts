// Addresses of tokens `balanceOf` storage slot.
// Used to set ERC20 account balance on fork
// internal details: https://docs.soliditylang.org/en/latest/internals/layout_in_storage.html#mappings-and-dynamic-arrays
import { AbiCoder, Addressable, keccak256, toBeHex } from 'ethers';
import { ethers } from 'hardhat';

export enum ArbMainnetERC20BalanceOfSlot {
  USDC = '0000000000000000000000000000000000000000000000000000000000000009',
  WETH = '0000000000000000000000000000000000000000000000000000000000000033',
  FRXETH = '0000000000000000000000000000000000000000000000000000000000000000',
  PTWEETH = '0000000000000000000000000000000000000000000000000000000000000000',
  WEETH = '0000000000000000000000000000000000000000000000000000000000000033',
}

// How to get: 1) decompile contract https://ethervm.io/decompile
// 2) find balanceOf function and slot
// or find slot in blockexplorer statechange  e.g. https://etherscan.io/tx/0xd3a83090d4e736aef85302e9835850d925c7d8da5180678fe440edc519966906#statechange
export enum EthereumMainnetERC20BalanceOfSlot {
  WETH = '0000000000000000000000000000000000000000000000000000000000000003',
  DAI = '0000000000000000000000000000000000000000000000000000000000000002',
  WBTC = '0000000000000000000000000000000000000000000000000000000000000000',
  USDC = '0000000000000000000000000000000000000000000000000000000000000009',
  SUSDE = '0000000000000000000000000000000000000000000000000000000000000004',
  PTSUSDE = '0000000000000000000000000000000000000000000000000000000000000000',
  PTSWINWSTETHS = '52c63247e1f47db19d5ce0460030c497f067ca4cebf71ba98eeadabe20bace00',
  INWSTETHS = '0000000000000000000000000000000000000000000000000000000000000065',
  WSTUSR = '52c63247e1f47db19d5ce0460030c497f067ca4cebf71ba98eeadabe20bace00',
  SDOLA = '0000000000000000000000000000000000000000000000000000000000000003',
  USR = '52c63247e1f47db19d5ce0460030c497f067ca4cebf71ba98eeadabe20bace00',
  DOLA = '0000000000000000000000000000000000000000000000000000000000000006',
}

function getAccountBalanceStorageSlot(account: string, tokenMappingSlot: string): string {
  return keccak256('0x' + account.slice(2).padStart(64, '0') + tokenMappingSlot);
}

export async function setTokenBalance(
  tokenAddress: string | Addressable,
  balanceOfSlotAddress: string,
  account: string,
  newBalance: bigint
) {
  const balanceOfStorageSlot = getAccountBalanceStorageSlot(account, balanceOfSlotAddress);

  await ethers.provider.send('hardhat_setStorageAt', [
    tokenAddress.toString(),
    balanceOfStorageSlot,
    toBeHex(newBalance, 32),
  ]);
}
