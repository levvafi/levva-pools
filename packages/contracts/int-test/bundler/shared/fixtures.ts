import { SignerWithAddress } from '@nomicfoundation/hardhat-ethers/signers';
import { ethers } from 'hardhat';
import { parseUnits, ZeroAddress } from 'ethers';
import { IERC20__factory, LevvaPendleBundler, LevvaPendleBundler__factory } from '../../../typechain-types';
import { LevvaFarmingPool, LevvaFarmingPool__factory } from '../../../typechain-types';
import { MarginlyParamsStruct } from '../../../typechain-types/contracts/LevvaFarmingPool';
import { MarginlyFactory__factory } from '../../../typechain-types';
import { MarginlyRouter__factory } from '../../../typechain-types';
import { PendleMarketAdapter__factory } from '../../../typechain-types';
import { PendleMarketOracle__factory } from '../../../typechain-types';
import { PENDLE_ROUTER, WETH9, PENDLE_ORACLE, WSTETH } from './registry';
import { PENDLE_PT_WSTETH_30_12_27, PENDLE_MARKET_WSTETH_30_12_27 } from './registry';
import { EthereumMainnetERC20BalanceOfSlot, setTokenBalance } from './tokens';
import { CallType } from '../../pool/utils/chain-ops';

export async function deployPendleBundlerWithPool(): Promise<{ bundler: LevvaPendleBundler; pool: LevvaFarmingPool }> {
  const [signer] = await ethers.getSigners();
  const pool = await deployPool(signer);
  const bundler = await deployPendleBundler(signer);
  return { pool, bundler };
}

async function deployPendleBundler(owner: SignerWithAddress): Promise<LevvaPendleBundler> {
  return new LevvaPendleBundler__factory(owner).deploy(PENDLE_ROUTER);
}

async function deployPool(owner: SignerWithAddress): Promise<LevvaFarmingPool> {
  const poolImplementation = await new LevvaFarmingPool__factory(owner).deploy();

  const pendleAdapter = await new PendleMarketAdapter__factory(owner).deploy([
    {
      pendleMarket: PENDLE_MARKET_WSTETH_30_12_27,
      slippage: 5,
      ptToken: PENDLE_PT_WSTETH_30_12_27,
      ibToken: WSTETH,
    },
  ]);
  const router = await new MarginlyRouter__factory(owner).deploy([
    { dexIndex: 0, adapter: await pendleAdapter.getAddress() },
  ]);

  const oracle = await new PendleMarketOracle__factory(owner).deploy(PENDLE_ORACLE);
  await oracle.setPair(WSTETH, PENDLE_PT_WSTETH_30_12_27, PENDLE_MARKET_WSTETH_30_12_27, 600, 10);

  const factory = await new MarginlyFactory__factory(owner).deploy(poolImplementation, router, owner, WETH9, owner);
  const poolAddress = await factory.createPool.staticCall(
    WSTETH,
    PENDLE_PT_WSTETH_30_12_27,
    oracle,
    0,
    defaultParams()
  );
  await factory.createPool(WSTETH, PENDLE_PT_WSTETH_30_12_27, oracle, 0, defaultParams());

  const pool = LevvaFarmingPool__factory.connect(poolAddress, owner.provider);

  const deposit = parseUnits('50', 18);
  await setTokenBalance(WSTETH, EthereumMainnetERC20BalanceOfSlot.WSTETH, owner.address, deposit);
  await IERC20__factory.connect(WSTETH, owner).approve(pool, deposit);
  await pool.connect(owner).execute(CallType.DepositQuote, deposit, 0, 0, false, ZeroAddress, 0);

  return pool;
}

function defaultParams(): MarginlyParamsStruct {
  return {
    interestRate: 0,
    fee: 0,
    maxLeverage: 20,
    swapFee: 0,
    mcSlippage: 10000,
    positionMinAmount: 1,
    quoteLimit: parseUnits('200', 18),
  };
}
