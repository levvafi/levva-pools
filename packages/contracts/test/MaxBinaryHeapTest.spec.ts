import { ethers } from 'hardhat';
import { expect } from 'chai';
import { loadFixture } from '@nomicfoundation/hardhat-network-helpers';
import { generateWallets } from './shared/utils';
import snapshotGasCost from 'optifat-snapshot-gas-cost';

describe('MaxBinaryHeapTest', () => {
  async function deployMaxBinaryHeapTestFixture() {
    // Contracts are deployed using the first signer/account by default
    const signers = await ethers.getSigners();
    const [owner, otherAccount] = await ethers.getSigners();

    const factory = await ethers.getContractFactory('MaxBinaryHeapTest');
    const contract = await factory.deploy();
    await contract.deployed();

    return { contract, owner, otherAccount, signers };
  }

  it('should create empty heap', async () => {
    const { contract } = await loadFixture(deployMaxBinaryHeapTestFixture);

    expect((await contract.getHeapLength()).toNumber()).to.equal(0);
  });

  it('should return false when peek root on empty heap', async () => {
    const { contract } = await loadFixture(deployMaxBinaryHeapTestFixture);

    const [peekSuccess, _] = await contract.getNodeByIndex(0);
    expect(peekSuccess).to.be.false;
  });

  it('should return isEmpty', async () => {
    const { contract, otherAccount } = await loadFixture(deployMaxBinaryHeapTestFixture);

    expect(await contract.isEmpty()).to.be.true;
    await contract.connect(otherAccount).add(10, otherAccount.address);
    expect(await contract.isEmpty()).to.be.false;
  });

  it('should add new item and rebuild tree', async () => {
    const { contract, owner, otherAccount } = await loadFixture(deployMaxBinaryHeapTestFixture);
    const item = 125;

    await snapshotGasCost(contract.connect(otherAccount).add(item, otherAccount.address));
    expect(await contract.getHeapLength()).to.equal(1);

    const [peekSuccess, root] = await contract.getNodeByIndex(0);
    expect(peekSuccess).to.be.true;
    expect(root.key).to.equal(item);
    expect(root.account).to.equal(otherAccount.address);

    const position = await contract.positions(otherAccount.address);
    expect(position.heapPosition).to.be.equal(1);
  });

  it('should return success=false when trying to get index of not existed in the heap account', async () => {
    const { contract, owner, otherAccount } = await loadFixture(deployMaxBinaryHeapTestFixture);
    const item = 125;

    await contract.connect(otherAccount).add(item, otherAccount.address);
  });

  it('should return success=false on empty heap', async () => {
    const { contract } = await loadFixture(deployMaxBinaryHeapTestFixture);
    const [success, node] = await contract.getNodeByIndex(0);
    expect(success).to.be.false;
  });

  it('should remove root item and rebuild tree', async () => {
    const { contract, owner, otherAccount } = await loadFixture(deployMaxBinaryHeapTestFixture);
    await contract.connect(otherAccount).add(1, otherAccount.address);
    await contract.connect(owner).add(2, owner.address);

    expect(await contract.getHeapLength()).to.equal(2);

    const [, rootItem] = await contract.connect(otherAccount).getNodeByIndex(0);
    await contract.connect(otherAccount).remove(0);
    expect(rootItem.key).to.equal(2);

    const removedPosition = await contract.positions(rootItem.account);
    expect(removedPosition.heapPosition).to.be.equal(0);

    expect(await contract.getHeapLength()).to.equal(1);

    const [, lastItem] = await contract.connect(otherAccount).getNodeByIndex(0);

    const lastItemPosition = await contract.positions(lastItem.account);
    expect(lastItemPosition.heapPosition).to.be.equal(1);
  });

  it('should remove item by index', async () => {
    /**
     * Create tree
     *        10
     *       /  \
     *      6   4
     *    / \
     *   3  5
     */
    const { contract, signers } = await loadFixture(deployMaxBinaryHeapTestFixture);
    await contract.connect(signers[0]).add(4, signers[0].address);
    await contract.connect(signers[1]).add(3, signers[1].address);
    await contract.connect(signers[2]).add(5, signers[2].address);
    await contract.connect(signers[3]).add(6, signers[3].address);
    await contract.connect(signers[4]).add(10, signers[4].address);

    await snapshotGasCost(contract.remove(1));

    let success, topItem;
    [success, topItem] = await contract.getNodeByIndex(0);
    await (await contract.remove(0)).wait();
    let position = await contract.positions(topItem.account);
    expect(success).to.be.true;
    expect(topItem.key).to.be.equal(10);
    expect(topItem.account).to.be.equal(signers[4].address);
    expect(position.heapPosition).to.be.equal(0);

    [success, topItem] = await contract.getNodeByIndex(0);
    await (await contract.remove(0)).wait();
    position = await contract.positions(topItem.account);
    expect(success).to.be.true;
    expect(topItem.key).to.be.equal(5);
    expect(topItem.account).to.be.equal(signers[2].address);
    expect(position.heapPosition).to.be.equal(0);

    [success, topItem] = await contract.getNodeByIndex(0);
    await (await contract.remove(0)).wait();
    position = await contract.positions(topItem.account);
    expect(success).to.be.true;
    expect(topItem.key).to.be.equal(4);
    expect(topItem.account).to.be.equal(signers[0].address);
    expect(position.heapPosition).to.be.equal(0);

    [success, topItem] = await contract.getNodeByIndex(0);
    await (await contract.remove(0)).wait();
    position = await contract.positions(topItem.account);
    expect(success).to.be.true;
    expect(topItem.key).to.be.equal(3);
    expect(topItem.account).to.be.equal(signers[1].address);
    expect(position.heapPosition).to.be.equal(0);

    [success, topItem] = await contract.getNodeByIndex(0);
    expect(success).to.be.false;
  });

  it('should remove from last to top', async () => {
    /**
     * Create tree
     *        10
     *       /  \
     *      6   4
     *    / \
     *   3  5
     */
    const { contract, signers } = await loadFixture(deployMaxBinaryHeapTestFixture);
    await contract.connect(signers[0]).add(4, signers[0].address);
    await contract.connect(signers[1]).add(3, signers[1].address);
    await contract.connect(signers[2]).add(5, signers[2].address);
    await contract.connect(signers[3]).add(6, signers[3].address);
    await contract.connect(signers[4]).add(10, signers[4].address);

    for (let i = 4; i >= 0; i--) {
      const [success, last] = await contract.getNodeByIndex(i);
      let position = await contract.positions(last.account);
      expect(position.heapPosition).to.be.eq(i + 1);

      await contract.remove(i);

      position = await contract.positions(last.account);
      expect(position.heapPosition).to.be.eq(0);
    }
  });

  it('should remove arbitary element ', async () => {
    /**
     * Create tree
     *        10
     *       /  \
     *      6   4
     *    / \
     *   3  5
     */
    const { contract, signers } = await loadFixture(deployMaxBinaryHeapTestFixture);
    await contract.connect(signers[0]).add(4, signers[0].address);
    await contract.connect(signers[1]).add(3, signers[1].address);
    await contract.connect(signers[2]).add(5, signers[2].address);
    await contract.connect(signers[3]).add(6, signers[3].address);
    await contract.connect(signers[4]).add(10, signers[4].address);

    const toRemove = 1;
    const [success, last] = await contract.getNodeByIndex(toRemove);
    let position = await contract.positions(last.account);
    expect(position.heapPosition).to.be.eq(toRemove + 1);

    await contract.remove(1);

    position = await contract.positions(last.account);
    expect(position.heapPosition).to.be.eq(0);
  });

  it('should remove element and update tree (heapifyUp)', async () => {
    /**
     * Create tree
     *          1008
     *        /      \
     *      1003     1006
     *     /   \      /
     *   1002  1000  1005
     */
    const { contract, signers } = await loadFixture(deployMaxBinaryHeapTestFixture);
    await contract.connect(signers[0]).add(1008, signers[0].address);
    await contract.connect(signers[1]).add(1003, signers[1].address);
    await contract.connect(signers[2]).add(1006, signers[2].address);
    await contract.connect(signers[3]).add(1002, signers[3].address);
    await contract.connect(signers[4]).add(1000, signers[4].address);
    await contract.connect(signers[5]).add(1005, signers[5].address);

    //remove signer[4] node
    const toRemove = 4;
    const [, node] = await contract.getNodeByIndex(toRemove);
    expect(node.account).to.be.eq(signers[4].address);
    expect(node.key).to.be.eq(1000);
    const position = await contract.positions(node.account);
    expect(position.heapPosition).to.be.eq(toRemove + 1);

    await contract.remove(toRemove);
    contract.updateByIndex(0, 1001);

    let prevKey = null;
    for (let i = 0; i < 5; i++) {
      const [, root] = await contract.getNodeByIndex(0);
      await contract.remove(0);

      if (prevKey) {
        expect(prevKey).to.be.greaterThanOrEqual(root.key);
      }

      prevKey = root.key;
    }
  });

  it('should remove element and update tree (heapifyDown)', async () => {
    /**
     * Create tree
     *          1008
     *        /      \
     *      1003     1006
     *     /   \      /
     *   1002  1000  1005
     */
    const { contract, signers } = await loadFixture(deployMaxBinaryHeapTestFixture);
    await contract.connect(signers[0]).add(1008, signers[0].address);
    await contract.connect(signers[1]).add(1003, signers[1].address);
    await contract.connect(signers[2]).add(1006, signers[2].address);
    await contract.connect(signers[3]).add(1002, signers[3].address);
    await contract.connect(signers[4]).add(1000, signers[4].address);
    await contract.connect(signers[5]).add(1005, signers[5].address);

    //remove root node
    const toRemove = 0;
    const [, node] = await contract.getNodeByIndex(toRemove);
    expect(node.account).to.be.eq(signers[0].address);
    expect(node.key).to.be.eq(1008);
    const position = await contract.positions(node.account);
    expect(position.heapPosition).to.be.eq(toRemove + 1);

    await contract.remove(toRemove);

    let prevKey = null;
    for (let i = 0; i < 5; i++) {
      const [, root] = await contract.getNodeByIndex(0);
      await contract.remove(0);

      if (prevKey) {
        expect(prevKey).to.be.greaterThanOrEqual(root.key);
      }

      prevKey = root.key;
    }
  });

  describe('Should update heap by index', () => {
    /**
     * Deploy contract and prepare heap
     *          50
     *        /    \
     *       40     30
     *      /  \
     *     20   10
     */
    async function prepareHeap() {
      const [owner, first, second, third, fourth] = await ethers.getSigners();

      const factory = await ethers.getContractFactory('MaxBinaryHeapTest');
      const contract = await factory.deploy();
      await contract.deployed();

      await contract.connect(owner).add(50, owner.address);
      await contract.connect(first).add(40, first.address);
      await contract.connect(second).add(30, second.address);
      await contract.connect(third).add(20, third.address);
      await contract.connect(fourth).add(10, fourth.address);

      expect(await contract.getHeapLength()).to.be.equal(5);

      {
        const [, node] = await contract.getNodeByIndex(0);
        const position = await contract.positions(node.account);
        expect(node.key).to.be.equal(50);
        expect(position.heapPosition).to.be.equal(1);
      }

      {
        const [, node] = await contract.getNodeByIndex(1);
        const position = await contract.positions(node.account);
        expect(node.key).to.be.equal(40);
        expect(position.heapPosition).to.be.equal(2);
      }

      {
        const [, node] = await contract.getNodeByIndex(2);
        const position = await contract.positions(node.account);
        expect(node.key).to.be.equal(30);
        expect(position.heapPosition).to.be.equal(3);
      }

      {
        const [, node] = await contract.getNodeByIndex(3);
        const position = await contract.positions(node.account);
        expect(node.key).to.be.equal(20);
        expect(position.heapPosition).to.be.equal(4);
      }

      {
        const [, node] = await contract.getNodeByIndex(4);
        const position = await contract.positions(node.account);
        expect(node.key).to.be.equal(10);
        expect(position.heapPosition).to.be.equal(5);
      }

      return { contract, owner, first, second, third, fourth };
    }

    it('without changing position', async () => {
      const { contract, owner } = await loadFixture(prepareHeap);
      await snapshotGasCost(contract.connect(owner).updateByIndex(0, 55));

      const [, rootNode] = await contract.getNodeByIndex(0);
      expect(rootNode.key.toNumber()).to.be.equal(55);
      expect(rootNode.account).to.be.equal(owner.address);
    });

    it('from middle to top', async () => {
      const { contract, owner, first } = await loadFixture(prepareHeap);
      await snapshotGasCost(contract.connect(owner).updateByIndex(1, 55));

      const [, rootNode] = await contract.getNodeByIndex(0);
      expect(rootNode.key.toNumber()).to.be.equal(55);
      expect(rootNode.account).to.be.equal(first.address);

      const [, firstNode] = await contract.getNodeByIndex(1);
      expect(firstNode.key.toNumber()).to.be.equal(50);
      expect(firstNode.account).to.be.equal(owner.address);
    });

    it('from middle to bottom', async () => {
      const { contract, first, third } = await loadFixture(prepareHeap);
      await snapshotGasCost(contract.connect(first).updateByIndex(1, 15));

      const [, middleNode] = await contract.getNodeByIndex(1);
      expect(middleNode.key.toNumber()).to.be.equal(20);
      expect(middleNode.account).to.be.equal(third.address);

      const [, bottomNode] = await contract.getNodeByIndex(3);
      expect(bottomNode.key.toNumber()).to.be.equal(15);
      expect(bottomNode.account).to.be.equal(first.address);
    });

    it('from top to bottom', async () => {
      const { contract, owner, first, third } = await loadFixture(prepareHeap);
      await snapshotGasCost(contract.connect(first).updateByIndex(0, 15));

      const [getTopNodeSuccess, topNode] = await contract.getNodeByIndex(0);
      expect(getTopNodeSuccess).to.be.true;
      expect(topNode.key.toNumber()).to.be.equal(40);
      expect(topNode.account).to.be.equal(first.address);

      const [getFirstNodeSuccess, firstNode] = await contract.getNodeByIndex(1);
      expect(getFirstNodeSuccess).to.be.true;
      expect(firstNode.key.toNumber()).to.be.equal(20);
      expect(firstNode.account).to.be.equal(third.address);

      const [getThirdNodeSuccess, thirdNode] = await contract.getNodeByIndex(3);
      expect(getThirdNodeSuccess).to.be.true;
      expect(thirdNode.key.toNumber()).to.be.equal(15);
      expect(thirdNode.account).to.be.equal(owner.address);
    });

    it('from bottom to top', async () => {
      const { contract, owner, first, third, fourth } = await loadFixture(prepareHeap);
      await snapshotGasCost(contract.connect(first).updateByIndex(4, 55));

      const [getTopNodeSuccess, topNode] = await contract.getNodeByIndex(0);
      expect(getTopNodeSuccess).to.be.true;
      expect(topNode.key.toNumber()).to.be.equal(55);
      expect(topNode.account).to.be.equal(fourth.address);

      const [getBottomNodeSuccess, bottomNode] = await contract.getNodeByIndex(4);
      expect(getBottomNodeSuccess).to.be.true;
      expect(bottomNode.key.toNumber()).to.be.equal(40);
      expect(bottomNode.account).to.be.equal(first.address);

      const [getFirstNodeSuccess, firstNode] = await contract.getNodeByIndex(1);
      expect(getFirstNodeSuccess).to.be.true;
      expect(firstNode.key.toNumber()).to.be.equal(50);
      expect(firstNode.account).to.be.equal(owner.address);
    });

    it('should update node account by index', async () => {
      const { contract, owner, first, third, fourth } = await loadFixture(prepareHeap);

      let [success, fourthNode] = await contract.getNodeByIndex(4);
      expect(fourthNode.account).to.be.equal(fourth.address);

      await snapshotGasCost(contract.connect(first).updateAccount(4, first.address));

      [success, fourthNode] = await contract.getNodeByIndex(4);
      expect(fourthNode.account).to.be.equal(first.address);
    });
  });

  it('should create max binary heap and remove items in right order', async () => {
    const { contract, otherAccount } = await loadFixture(deployMaxBinaryHeapTestFixture);
    const wallets = await generateWallets(100);
    const heapLength = 100;

    expect(await contract.getHeapLength()).to.equal(0);

    for (let i = 0; i < heapLength; i++) {
      const value = i + 1;

      if (i == heapLength - 1) {
        //measure only last call
        await snapshotGasCost(contract.connect(otherAccount).add(value, wallets[i].address));
      } else {
        await contract.connect(otherAccount).add(value, wallets[i].address);
      }
    }

    expect(await contract.getHeapLength()).to.equal(heapLength);

    let prevKey;
    for (let i = 0; i < heapLength; i++) {
      const [_, topItem] = await contract.getNodeByIndex(0);
      const itemKey = topItem.key.toNumber();

      if (i == 0) {
        //measure only first call
        await snapshotGasCost(contract.connect(otherAccount).remove(0));
      } else {
        await contract.connect(otherAccount).remove(0);
      }

      // skip the very first check to ensure all next elements strict greater than prev
      if (i > 0) {
        expect(prevKey).to.be.greaterThan(itemKey);
      }
      prevKey = itemKey;
    }

    expect(await contract.getHeapLength()).to.equal(0);
  });
});
