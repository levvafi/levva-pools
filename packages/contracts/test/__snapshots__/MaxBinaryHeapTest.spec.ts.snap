// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MaxBinaryHeapTest Should update heap by index from bottom to top 1`] = `
Object {
  "calldataByteLength": 68,
  "gasUsed": 58459,
}
`;

exports[`MaxBinaryHeapTest Should update heap by index from middle to bottom 1`] = `
Object {
  "calldataByteLength": 68,
  "gasUsed": 49383,
}
`;

exports[`MaxBinaryHeapTest Should update heap by index from middle to top 1`] = `
Object {
  "calldataByteLength": 68,
  "gasUsed": 46337,
}
`;

exports[`MaxBinaryHeapTest Should update heap by index from top to bottom 1`] = `
Object {
  "calldataByteLength": 68,
  "gasUsed": 64095,
}
`;

exports[`MaxBinaryHeapTest Should update heap by index should update node account by index 1`] = `
Object {
  "calldataByteLength": 68,
  "gasUsed": 27231,
}
`;

exports[`MaxBinaryHeapTest Should update heap by index without changing position 1`] = `
Object {
  "calldataByteLength": 68,
  "gasUsed": 29403,
}
`;

exports[`MaxBinaryHeapTest should add new item and rebuild tree 1`] = `
Object {
  "calldataByteLength": 68,
  "gasUsed": 89186,
}
`;

exports[`MaxBinaryHeapTest should create max binary heap and remove items in right order 1`] = `
Object {
  "calldataByteLength": 68,
  "gasUsed": 144819,
}
`;

exports[`MaxBinaryHeapTest should create max binary heap and remove items in right order 2`] = `
Object {
  "calldataByteLength": 36,
  "gasUsed": 112862,
}
`;

exports[`MaxBinaryHeapTest should remove item by index 1`] = `
Object {
  "calldataByteLength": 36,
  "gasUsed": 41630,
}
`;
